### PR TITLE
all: add side-channel logf communication via magic args

### PR DESCRIPTION
**Pull request by josharian**
I suggest reviewing this commit by commit.

This is perhaps both too cute and too clever, but:

* it is a clear, extensible, re-usable approach to the problem of cross-layer logf communication
* I've not come up with any alternatives that aren't abysmal

If we decide to move forward with this, the next steps are:

* see about eliminating logger.StdLogger wherever possible
* consider using logger.NoRateLimit instead of the magic prefixes
* consider applying logger.RateLimitContext to places where we have very plain format strings
* consider whether to use this approach for other things, like verbosity

### CODE REVIEW

File: types/logger/logger.go [hunk 1]
```diff
@@ -92,10 +97,20 @@ func RateLimitedFn(logf Logf, f time.Duration, burst int, maxCache int) Logf {
...
+	// judge decides the fate of a log request and returns the string that should be used
+	// to describe the format when the verdict is warn.
+	judge := func(format string, args ...interface{}) (v verdict, warnFormat string) {
+		for _, arg := range args {
+			switch arg := arg.(type) {
+			case noRateLimit:
+				return allow, ""
+			}
+		}
+		format = noopFormatRemover.Replace(format)
```
Comment (bradfitz) [inline types/logger/logger.go:116]: can we relay this call until it's needed on line 139? I don't think the line 112 cares.

File: types/logger/logger.go [hunk 2]
```diff
@@ -216,3 +231,35 @@ func LogfCloser(logf Logf) (newLogf Logf, close func()) {
...
+const noopFormat = "%+5.2L"
+
+var noopFormatRemover = strings.NewReplacer(noopFormat, "")
+
+// noopFormatter is a type that generates nothing when printing using fmt.Sprintf.
+// It may be embedded in types for internal-use args, so that, which used
+// in correspondence with noopFormat, they have no impact on the actual log output.
+type noopFormatter struct{}
+
+func (f noopFormatter) Format(fmt.State, rune) {}
```
Comment (bradfitz) [inline types/logger/logger.go:258]: drop receiver name

File: types/logger/logger.go [hunk 3]
```diff
@@ -216,3 +231,35 @@ func LogfCloser(logf Logf) (newLogf Logf, close func()) {
...
+		logf(format+noopFormat, xargs...)
+	}
+}
+
+// NoRateLimit removes rate limiting for logf.
+func NoRateLimit(logf Logf) Logf {
+	return logfWithExtra(logf, noRateLimit{})
+}
+
+type noRateLimit struct {
```
Comment (bradfitz) [inline types/logger/logger.go:274]: comment

File: types/logger/logger.go [hunk 4]
```diff
@@ -114,6 +121,10 @@ func RateLimitedFn(logf Logf, f time.Duration, burst int, maxCache int) Logf {
 			}
 		}
 
+		if len(contexts) > 0 {
+			format += " (rate-limit-context:" + strings.Join(contexts, ",") + ")"
```
Comment (bradfitz) [inline types/logger/logger.go:125]: this seems like a lot of allocs per Logf.

File: types/logger/logger.go [hunk 5]
```diff
@@ -284,3 +284,47 @@ type rateLimitContext struct {
 	noopFormatter
 	context string
 }
+
+// ApplyPostProcess works with PostProcess to allow
```
Comment (bradfitz) [inline types/logger/logger.go:288]: I read this godoc several times but still don't get what it does from its description.

Why is this exported if it seems like just an implementation detail of PostProcess which has understandable docs?

File: types/logger/logger.go (hunk 1 above)
Comment (josharian) [reply types/logger/logger.go:116]: It's needed on line 119, but delayed past the `rateFreePrefix` loop.

Comment (bradfitz) [reply types/logger/logger.go:116]: why do we need it on line 119? it's fine if there's a %5.2L gibberish in there, as users don't see it.

I'd rather not allocate on each rate-limited log print.

File: types/logger/logger.go (hunk 4 above)
Comment (josharian) [reply types/logger/logger.go:125]: All of this is really expensive. This line is only a single alloc. But there are also two allocs in logfExtra (a string concatenation and a slice alloc), and two allocs from the strings.Replacer call.

I'll add a benchmark suite in a final commit. Results (per log entry):

```
BenchmarkRateLimiter-8          	 6015066	       194 ns/op	       0 B/op	       0 allocs/op
BenchmarkNoRateLimit-8          	10655953	       107 ns/op	      96 B/op	       2 allocs/op
BenchmarkRateLimiterContext-8   	 2654502	       451 ns/op	     176 B/op	       5 allocs/op
BenchmarkStdLogger-8            	 4168108	       286 ns/op	      80 B/op	       3 allocs/op
BenchmarkPostProcess-8          	 2268715	       529 ns/op	     136 B/op	       8 allocs/op
```

File: types/logger/logger.go (hunk 5 above)
Comment (josharian) [reply types/logger/logger.go:288]: I rewrote. Please let me know if it remains unclear.

File: types/logger/logger.go (hunk 1 above)
Comment (josharian) [reply types/logger/logger.go]: Moved. Did it in a follow-up, because the rebase got ugly.

**Comment by josharian** [comment]
To be clear, I'm not wedded to this. It's the best idea I've had to deal with this issue working within the constraints of the existing logging interface, but "best idea" doesn't imply "good idea".

**Comment by josharian** [comment]
This was exploratory.

**Pull request closed without merging**